File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int dp[1001][1001];
    int go(string &s1,string &s2,int a,int b){
        if(dp[a][b]!=-1) return dp[a][b];
        int &ans = dp[a][b];
        if(a==0 or b==0) return dp[a][b] =0;
        if(s1[a-1]==s2[b-1]) return dp[a][b]=1+go(s1,s2,a-1,b-1);
        return dp[a][b] = max(go(s1,s2,a,b-1),go(s1,s2,a-1,b));
    }
    int longestCommonSubsequence(string text1, string text2) {
        memset(dp,-1,sizeof(dp));
        return go(text1,text2,text1.size(),text2.size());
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, t1: str, t2: str) -> int:
        m=len(t1)
        n=len(t2)
        # dp = [[0]*(m+1)]*(n+1);
        # dp = [[0]*(m+1) for x in range(n+1)]
        dp = [[0 for y in range (m+1) ] for x in range(n+1) ]
        for i in range(1,n+1):
            for j in range (1,m+1):
                if(t2[i-1]==t1[j-1]):
                    dp[i][j]= dp[i-1][j-1]+1;
                else :
                    dp[i][j] = max( dp[i-1][j],dp[i][j-1]);
        return dp[n][m]